### PR TITLE
feat(project): cache improvements

### PR DESCRIPTION
**Pull request by eloy-encord**
Move data management load to Project class in order to encapsulate all read/save/update data from an Encord's project in the platform to its local equivalent.

Now, further improvements on cache load/dump can be written only in the Project class and remove the complications on how to handle it on other code sections.

### CODE REVIEW

File: src/encord_active/lib/common/project.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        subset_size: Optional[int] = None,
+        encord_project: Optional[EncordProject] = None,
+        **kwargs,
+    ) -> Project:
+        """
+        Fetches data either from cache or from the Encord project if it's available.
+
+        :param cache_dir: The root directory of the project.
+        :param subset_size: The number of label rows to fetch. None means all.
+        :param encord_project: An encord project for fetching data. If not provided, will only use cached files.
```
Comment (Encord-davids) [inline src/encord_active/lib/common/project.py:50]: [curious]🤔 

just to make sure I understand this, is the intended usage something like

```python
# setup project
project = Project.read("some/path", encord_project=EncordProject("whatever"))
project.save()

# in some other file, just use local project
project = Project.read("some/path")
```

File: src/encord_active/lib/common/project.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        label_row_meta_file_path.write_text(json.dumps(label_row_meta, indent=2))  # , encoding="utf-8")
+
+        # store label rows and their images
+        label_rows = download_all_label_rows(encord_project, cache_dir=project_dir)  # todo no need to output all data
+        image_paths = download_all_images(label_rows, cache_dir=project_dir)  # todo no need to output all data
+
+        return Project(project_dir)
+
+    # unused method (prototype)
+    def save(self, cache_dir: Path) -> None:
```
Comment (Encord-davids) [inline src/encord_active/lib/common/project.py:124]: I think this is kind of useless at this point since we don't really change the local project and this isn't used by the `from_encord_project` which seems to replicate this logic.

I don't want to delay this PR event more, but I think we should iterate over this in the future.

I think that separating what's in memory and what's on disc can be useful here. We should be able to create a `Project` from an encord project without saving anything and only when `save` is called, we dump everything to disc. 

Also, in the future, we might want the ability to have a flow of creating a project from the UI and for that, we will also need to be able to instantiate a `Project` with values rather than a path. (loading from disc would be in a `load` method for example)

File: src/encord_active/lib/model_predictions/prediction_writer.py [hunk 3]
```diff
@@ -173,7 +173,8 @@ def __init__(
         self.object_lookup = {o["featureNodeHash"]: o for o in self.project.ontology["objects"]}
 
         logger.info("Fetching project label rows to be able to match predictions.")
-        self.label_rows = prepare_data(cache_dir, project=project, **kwargs).label_rows
+        # todo check if it is really necessary to read from project from Encord (probably should be local storage)
+        self.label_rows = Project.from_encord_project(cache_dir, project).label_rows
```
Comment (Encord-davids) [inline src/encord_active/lib/model_predictions/prediction_writer.py:177]: pretty sure that is correct since to import predictions we need a local project

File: src/encord_active/lib/common/project.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        label_row_meta_file_path = project_dir / "label_row_meta.json"
+        if not label_row_meta_file_path.exists():
+            raise FileNotFoundError(f"Expected file `label_row_meta.json` at {project_dir}")
+        self.label_row_meta: Dict[str, LabelRowMetadata] = {}
+        for lr_hash, lr_meta in itertools.islice(json.loads(label_row_meta_file_path.read_text()).items(), subset_size):
+            self.label_row_meta[lr_hash] = LabelRowMetadata.from_dict(lr_meta)
+
+        # read label rows and their images
+        self.label_rows: Dict[str, LabelRow] = {}
+        self.image_paths: Dict[str, List[Path]] = {}
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:65]: The way it works right now is that the iterator, for example, will take the stem of that path as the data unit hash so it will do the comparison there. We probably should take that off the iterator and convert `List[Path]` to a dict of data unit hash to image path.

File: src/encord_active/lib/common/project.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        """
+        Fetch project's data from local storage.
+
+        :param project_dir: The root directory of the project.
+        :param subset_size: The number of label rows to fetch. If None then all label rows are fetched.
+        """
+        if not project_dir.exists():
+            raise FileNotFoundError("`project_dir` must point to an existing directory")
+        if not project_dir.is_dir():
+            raise NotADirectoryError("`project_dir` must point to an existing directory")
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:40]: I stated here that way to signal the user what is exactly the problem with the provided path: "there is no such directory" or "pointing to wrong place". More or less a shortcut for the user.

File: src/encord_active/lib/common/project.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        # read label rows and their images
+        self.label_rows: Dict[str, LabelRow] = {}
+        self.image_paths: Dict[str, List[Path]] = {}
+        for lr_hash in self.label_row_meta.keys():
+            lr_file_path = project_dir / "data" / lr_hash / "label_row.json"
+            lr_images_dir = project_dir / "data" / lr_hash / "images"
+            if not lr_file_path.exists():  # todo log this issue
+                continue
+            if not lr_images_dir.exists() or not lr_images_dir.is_dir():  # todo log this issue
+                continue
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:72]: good idea

**Review by frederik-encord — APPROVED** [review]

**Comment by frederik-encord** [comment]
Woops. Approved by accident. I rerequested reveiew to allow David's comments to be taken care of.

File: src/encord_active/lib/model_predictions/prediction_writer.py (hunk 3 above)
Comment (frederik-encord) [reply src/encord_active/lib/model_predictions/prediction_writer.py:177]: I 100% agree.

File: src/encord_active/lib/common/project.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        Construct a Project from Encord platform.
+
+        :param project_dir: The root directory of the project.
+        :param encord_project: Encord project from where the data is fetched.
+        :return:
+        """
+        if not project_dir.exists():
+            raise FileNotFoundError("`project_dir` must point to an existing directory")
+        if not project_dir.is_dir():
+            raise NotADirectoryError("`project_dir` must point to an existing directory")
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:88]: sure, that would do the trick and it's more reasonable than forcing the user to create a folder on its own 👍

File: src/encord_active/lib/common/project.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+        # todo enforce clean up when we are sure it won't impact performance in other sections (like PredictionWriter)
+        # also don't forget to add `from shutil import rmtree` at the top (pylint tags it as unused right now)
+        # clean project_dir content
+        # for path in project_dir.iterdir():
+        #     if path.is_file():
+        #         path.unlink()
+        #     elif path.is_dir():
+        #         rmtree(path)
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:97]: The plan of this constructor is to actually update everything inside the data folder so we want to overwrite. As it is written now it will do a clean sweep of the folder, including metrics and whatnots, and we surely don't want that but it's a start 😄. Is not very clear what should remain here because metrics cvs files will become obsolete also and in the future we will want to run the metrics only in the data that changed.

To sum up, how to correctly update the data and metrics it's a topic on its own. Also, the value proposition of current project structure is to limit (get rid of) optional parameters like `override=False`.

File: src/encord_active/lib/common/project.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+            json.loads((project_dir / "ontology.json").read_text())
+        )
+
+        # read label rows' metadata
+        label_row_meta_file_path = project_dir / "label_row_meta.json"
+        if not label_row_meta_file_path.exists():
+            raise FileNotFoundError(f"Expected file `label_row_meta.json` at {project_dir}")
+        self.label_row_meta: Dict[str, LabelRowMetadata] = {}
+        for lr_hash, lr_meta in itertools.islice(json.loads(label_row_meta_file_path.read_text()).items(), subset_size):
+            self.label_row_meta[lr_hash] = LabelRowMetadata.from_dict(lr_meta)
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:61]: We can do better (more concise) 😃

File: src/encord_active/lib/common/project.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        #         rmtree(path)
+
+        # store project's metadata
+        project_meta = {
+            "project_title": encord_project.title,
+            "project_description": encord_project.description,
+            "project_hash": encord_project.project_hash,
+        }
+        project_meta_file_path = project_dir / "project_meta.yaml"
+        project_meta_file_path.write_text(yaml.dump(project_meta))  # , encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:101]: ups, these comments were meant to be local only

[the remark this answers is not in the document]
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:101]: better to stick to UTF-8 everywhere

File: src/encord_active/lib/common/project.py (hunk 2 above)
Comment (eloy-encord) [reply src/encord_active/lib/common/project.py:122]: I like the idea of separating memory and disk stuff but it will only work for prototyping, not production use cases as you would have to handle a massive amount of images/videos or to always have an open connection with the server that provides the data which adds way more time overload.

File: src/encord_active/lib/common/project.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/lib/common/project.py:40]: Yeah, that's what I assumed, but it doesn't make sense to raise a `FileNotFoundError` when we don't find the directory IMO 🤷 

Also, if we still consider them different errors, we should make the message more concrete like "`project_dir` doesn't exist" and "`project_dir` is pointing to a file instead of a directory"

File: src/encord_active/lib/common/project.py (hunk 2 above)
Comment (Encord-davids) [reply src/encord_active/lib/common/project.py]: Sorry, I wasn't clear, I don't think we should be holding all images in memory, just the data. We can fetch the media on demand or require an explicit download before media can be displayed, but the interim in-memory object could still be valuable.

File: src/encord_active/lib/common/project.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/lib/common/project.py:68]: Yes that sound much better 👍

File: src/encord_active/lib/common/project.py [hunk 11]
```diff
@@ -27,52 +27,64 @@
...
+        self.project_dir: Path
+        self.project_meta: Dict[str, str]
+        self.project_hash: str
+        self.ontology: OntologyStructure
+        self.label_row_meta: Dict[str, LabelRowMetadata]
+        self.label_rows: Dict[str, LabelRow]
+        self.image_paths: Dict[str, List[Path]]
+
+    @classmethod
+    def load(cls, project_dir: Path, subset_size: Optional[int] = None):
```
Comment (Encord-davids) [inline src/encord_active/lib/common/project.py:40]: Maybe like this instead, wdyt?

```suggestion
    def __init__(self, project_dir: Optional[Path] = None):
        self.project_dir = project_dir
        self.project_meta: Dict[str, str]
        self.project_hash: str
        self.ontology: OntologyStructure
        self.label_row_meta: Dict[str, LabelRowMetadata]
        self.label_rows: Dict[str, LabelRow]
        self.image_paths: Dict[str, List[Path]]

    def load(self, subset_size: Optional[int] = None):
```

This way you don't need `fake_project` and you can do `Project("/some/path").load()`

**Review by Encord-davids — DISMISSED** [review]

**Review by Encord-davids — APPROVED** [review]
